nomos-mix/message/src/layered_cipher.rs

```diff
@@ -0,0 +1,331 @@
+        let total_data = concat_bytes(&[
+            &param.data.to_bytes(),
+            next_mac.as_bytes(),
+            // Truncate last bytes for the length-preserved decryption later.
+            // They will be restored by a filler during the decryption process.
+            &next_encrypted_data[..next_encrypted_data.len() - Self::single_layer_size()],
+        ]);
+
+        // Encrypt the concatenated bytes, and compute MAC.
+        let mut encrypted = total_data;
+        self.apply_streamcipher(
+            &mut encrypted,
+            &param.key.stream_cipher_key,
+            StreamCipherOption::FromFront,
+        );
+        let mac = Self::compute_mac(&param.key.integrity_mac_key, &encrypted);
+
+        assert_eq!(encrypted.len(), self.total_size());
+        Ok((encrypted, mac))
+    }
+
+    /// Build the last layer of encryption.
+    /// The output has the same size as [`Self::total_size`] by using fillers,
+    /// even though it doesn't wrap any subsequent layer.
+    /// This is for the length-preserved decryption.
+    fn build_last_layer(
+        &self,
+        params: &[EncryptionParam<D>],
+    ) -> Result<(Vec<u8>, HeaderIntegrityMac)> {
+        let last_param = params.last().ok_or(Error::InvalidParams)?;
+
+        // Build fillers that will be appended to the last data.
+        // The number of fillers must be the same as the number of intermediate layers
+        // that will be decrypted later.
+        let fillers = self.build_fillers(&params[..params.len() - 1]);
+        // Random bytes is used to fill the space between data and fillers.
+        let random_bytes = random_bytes(self.total_size() - D::size() - fillers.len());
+
+        // First, concat the data and the random bytes, and encrypt it.
+        let total_data_without_fillers =
+            concat_bytes(&[&last_param.data.to_bytes(), &random_bytes]);
+        let mut encrypted = total_data_without_fillers;
+        self.apply_streamcipher(
+            &mut encrypted,
+            &last_param.key.stream_cipher_key,
+            StreamCipherOption::FromFront,
+        );
+
+        // Append fillers to the encrypted bytes, and compute MAC.
+        encrypted.extend(fillers);
+        let mac = Self::compute_mac(&last_param.key.integrity_mac_key, &encrypted);
+
+        assert_eq!(encrypted.len(), self.total_size());
+        Ok((encrypted, mac))
+    }
+
+    /// Build as many fillers as the number of keys provided.
+    /// Fillers are encrypted in accumulated manner by keys.
+    fn build_fillers(&self, params: &[EncryptionParam<D>]) -> Vec<u8> {
+        params
+            .iter()
+            .map(|param| &param.key.stream_cipher_key)
+            .fold(Vec::new(), |mut fillers, key| {
+                fillers.extend(vec![0u8; Self::single_layer_size()]);
+                self.apply_streamcipher(&mut fillers, key, StreamCipherOption::FromBack);
+                fillers
+            })
```

Comment:
Nitpick: You can probably allocate the full vec beforehand and just call `apply_stream_cipher` to the region you need?

Reply:
Right. Thanks. [3d32942](https://github.com/logos-co/nomos-node/pull/897/commits/3d32942725a2d769e6a928b74426c97d287f11ab)